File: tw-gradcs-linear-algebra/04-orthogonality-and-least-squares/05-qr-decomposition/cpp/qr_decomposition.cpp
/**
 * QR 分解 (QR Decomposition)
 *
 * 編譯：g++ -std=c++17 -O2 qr_decomposition.cpp -o qr_decomposition
 * 執行：./qr_decomposition
 */

#include <iostream>
#include <vector>
#include <cmath>
#include <iomanip>

using namespace std;

void printSeparator(const string& title) {
    cout << endl;
    cout << string(60, '=') << endl;
    cout << title << endl;
    cout << string(60, '=') << endl;
}

void printVector(const string& name, const vector<double>& v) {
    cout << name << " = [";
    for (size_t i = 0; i < v.size(); i++) {
        cout << fixed << setprecision(4) << v[i];
        if (i < v.size() - 1) cout << ", ";
    }
    cout << "]" << endl;
}

void printMatrix(const string& name, const vector<vector<double>>& M) {
    cout << name << " =" << endl;
    for (const auto& row : M) {
        cout << "  [";
        for (size_t i = 0; i < row.size(); i++) {
            cout << fixed << setprecision(4) << setw(8) << row[i];
            if (i < row.size() - 1) cout << ", ";
        }
        cout << "]" << endl;
    }
}

// 基本向量運算
double dotProduct(const vector<double>& x, const vector<double>& y) {
    double result = 0.0;
    for (size_t i = 0; i < x.size(); i++) {
        result += x[i] * y[i];
    }
    return result;
}

double vectorNorm(const vector<double>& x) {
    return sqrt(dotProduct(x, x));
}

vector<double> scalarMultiply(double c, const vector<double>& x) {
    vector<double> result(x.size());
    for (size_t i = 0; i < x.size(); i++) {
        result[i] = c * x[i];
    }
    return result;
}

vector<double> vectorSubtract(const vector<double>& x, const vector<double>& y) {
    vector<double> result(x.size());
    for (size_t i = 0; i < x.size(); i++) {
        result[i] = x[i] - y[i];
    }
    return result;
}

// 取得矩陣的第 j 行（column）
vector<double> getColumn(const vector<vector<double>>& A, int j) {
    vector<double> col(A.size());
    for (size_t i = 0; i < A.size(); i++) {
        col[i] = A[i][j];
    }
    return col;
}

// Gram-Schmidt QR 分解
pair<vector<vector<double>>, vector<vector<double>>>
qrDecomposition(const vector<vector<double>>& A) {
    int m = A.size();
    int n = A[0].size();

    // Q: m×n, R: n×n
    vector<vector<double>> Q(m, vector<double>(n, 0.0));
    vector<vector<double>> R(n, vector<double>(n, 0.0));

    for (int j = 0; j < n; j++) {
        // 取得 A 的第 j 行
        vector<double> v = getColumn(A, j);

        // 減去前面所有 q 向量的投影
        for (int i = 0; i < j; i++) {
            vector<double> qi = getColumn(Q, i);
            R[i][j] = dotProduct(qi, getColumn(A, j));
            vector<double> proj = scalarMultiply(R[i][j], qi);
            v = vectorSubtract(v, proj);
        }

        // 標準化
        R[j][j] = vectorNorm(v);

        if (R[j][j] > 1e-10) {
            for (int i = 0; i < m; i++) {
                Q[i][j] = v[i] / R[j][j];
            }
        }
    }

    return {Q, R};
}

// 回代法解上三角方程組 Rx = b
vector<double> solveUpperTriangular(const vector<vector<double>>& R,
                                     const vector<double>& b) {
    int n = b.size();
    vector<double> x(n, 0.0);

    for (int i = n - 1; i >= 0; i--) {
        x[i] = b[i];
        for (int j = i + 1; j < n; j++) {
            x[i] -= R[i][j] * x[j];
        }
        x[i] /= R[i][i];
    }

    return x;
}

// 用 QR 分解解最小平方問題
vector<double> qrLeastSquares(const vector<vector<double>>& A,
                               const vector<double>& b) {
    auto [Q, R] = qrDecomposition(A);

    // 計算 Qᵀb
    int n = Q[0].size();
    vector<double> Qt_b(n, 0.0);
    for (int j = 0; j < n; j++) {
        vector<double> qj = getColumn(Q, j);
        Qt_b[j] = dotProduct(qj, b);
    }

    // 解 Rx = Qᵀb
    return solveUpperTriangular(R, Qt_b);
}

// 矩陣乘法
vector<vector<double>> matrixMultiply(const vector<vector<double>>& A,
                                       const vector<vector<double>>& B) {
    int m = A.size();
    int k = B.size();
    int n = B[0].size();

    vector<vector<double>> result(m, vector<double>(n, 0.0));
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            for (int p = 0; p < k; p++) {
                result[i][j] += A[i][p] * B[p][j];
            }
        }
    }
    return result;
}

// 矩陣轉置
vector<vector<double>> transpose(const vector<vector<double>>& A) {
    int m = A.size();
    int n = A[0].size();
    vector<vector<double>> result(n, vector<double>(m));
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            result[j][i] = A[i][j];
        }
    }
    return result;
}

int main() {
    printSeparator("QR 分解示範 (C++)");

    // ========================================
    // 1. 基本 QR 分解
    // ========================================
    printSeparator("1. 基本 QR 分解");

    vector<vector<double>> A = {
        {1.0, 1.0},
        {1.0, 0.0},
        {0.0, 1.0}
    };

    cout << "輸入矩陣 A：" << endl;
    printMatrix("A", A);

    auto [Q, R] = qrDecomposition(A);

    cout << "\nQR 分解結果：" << endl;
    printMatrix("Q", Q);
    printMatrix("\nR", R);

    // 驗證 QᵀQ = I
    auto QT = transpose(Q);
    auto QTQ = matrixMultiply(QT, Q);
    cout << "\n驗證 QᵀQ = I：" << endl;
    printMatrix("QᵀQ", QTQ);

    // 驗證 A = QR
    auto QR_result = matrixMultiply(Q, R);
    cout << "\n驗證 A = QR：" << endl;
    printMatrix("QR", QR_result);

    // ========================================
    // 2. 用 QR 解最小平方
    // ========================================
    printSeparator("2. 用 QR 解最小平方");

    // 數據
    vector<double> t = {0.0, 1.0, 2.0};
    vector<double> b = {1.0, 3.0, 4.0};

    cout << "數據點：" << endl;
    for (size_t i = 0; i < t.size(); i++) {
        cout << "  (" << t[i] << ", " << b[i] << ")" << endl;
    }

    // 設計矩陣
    vector<vector<double>> A_ls(t.size(), vector<double>(2));
    for (size_t i = 0; i < t.size(); i++) {
        A_ls[i][0] = 1.0;
        A_ls[i][1] = t[i];
    }

    cout << "\n設計矩陣 A：" << endl;
    printMatrix("A", A_ls);
    printVector("觀測值 b", b);

    // QR 分解
    auto [Q_ls, R_ls] = qrDecomposition(A_ls);
    printMatrix("\nQ", Q_ls);
    printMatrix("R", R_ls);

    // 解最小平方
    vector<double> x = qrLeastSquares(A_ls, b);
    printVector("\n解 x", x);

    cout << "\n最佳直線：y = " << fixed << setprecision(4)
         << x[0] << " + " << x[1] << "t" << endl;

    // ========================================
    // 3. 3×3 矩陣的 QR 分解
    // ========================================
    printSeparator("3. 3×3 矩陣的 QR 分解");

    vector<vector<double>> A2 = {
        {1.0, 1.0, 0.0},
        {1.0, 0.0, 1.0},
        {0.0, 1.0, 1.0}
    };

    cout << "輸入矩陣 A：" << endl;
    printMatrix("A", A2);

    auto [Q2, R2] = qrDecomposition(A2);

    cout << "\nQR 分解結果：" << endl;
    printMatrix("Q", Q2);
    printMatrix("\nR", R2);

    // 總結
    printSeparator("總結");
    cout << R"(
QR 分解核心：

1. A = QR
   - Q: 標準正交矩陣 (QᵀQ = I)
   - R: 上三角矩陣

2. Gram-Schmidt 演算法：
   - 對 A 的行向量正交化得到 Q
   - R 的元素是投影係數

3. 用 QR 解最小平方：
   min ‖Ax - b‖²
   → Rx = Qᵀb

4. 優勢：
   - 比正規方程更穩定
   - 避免計算 AᵀA
)" << endl;

    cout << string(60, '=') << endl;
    cout << "示範完成！" << endl;
    cout << string(60, '=') << endl;

    return 0;
}

File: tw-gradcs-linear-algebra/04-orthogonality-and-least-squares/05-qr-decomposition/javascript/qr_decomposition.js
/**
 * QR 分解 (QR Decomposition)
 *
 * 執行：node qr_decomposition.js
 */

function printSeparator(title) {
    console.log();
    console.log('='.repeat(60));
    console.log(title);
    console.log('='.repeat(60));
}

function printVector(name, v) {
    const formatted = v.map(x => x.toFixed(4)).join(', ');
    console.log(`${name} = [${formatted}]`);
}

function printMatrix(name, M) {
    console.log(`${name} =`);
    for (const row of M) {
        const formatted = row.map(x => x.toFixed(4).padStart(8)).join(', ');
        console.log(`  [${formatted}]`);
    }
}

// 基本向量運算
function dotProduct(x, y) {
    return x.reduce((sum, xi, i) => sum + xi * y[i], 0);
}

function vectorNorm(x) {
    return Math.sqrt(dotProduct(x, x));
}

function scalarMultiply(c, x) {
    return x.map(xi => c * xi);
}

function vectorSubtract(x, y) {
    return x.map((xi, i) => xi - y[i]);
}

// 取得矩陣的第 j 行（column）
function getColumn(A, j) {
    return A.map(row => row[j]);
}

// Gram-Schmidt QR 分解
function qrDecomposition(A) {
    const m = A.length;
    const n = A[0].length;

    // Q: m×n, R: n×n
    const Q = Array.from({length: m}, () => Array(n).fill(0));
    const R = Array.from({length: n}, () => Array(n).fill(0));

    for (let j = 0; j < n; j++) {
        // 取得 A 的第 j 行
        let v = getColumn(A, j);

        // 減去前面所有 q 向量的投影
        for (let i = 0; i < j; i++) {
            const qi = getColumn(Q, i);
            R[i][j] = dotProduct(qi, getColumn(A, j));
            const proj = scalarMultiply(R[i][j], qi);
            v = vectorSubtract(v, proj);
        }

        // 標準化
        R[j][j] = vectorNorm(v);

        if (R[j][j] > 1e-10) {
            for (let i = 0; i < m; i++) {
                Q[i][j] = v[i] / R[j][j];
            }
        }
    }

    return [Q, R];
}

// 回代法解上三角方程組 Rx = b
function solveUpperTriangular(R, b) {
    const n = b.length;
    const x = Array(n).fill(0);

    for (let i = n - 1; i >= 0; i--) {
        x[i] = b[i];
        for (let j = i + 1; j < n; j++) {
            x[i] -= R[i][j] * x[j];
        }
        x[i] /= R[i][i];
    }

    return x;
}

// 用 QR 分解解最小平方問題
function qrLeastSquares(A, b) {
    const [Q, R] = qrDecomposition(A);

    // 計算 Qᵀb
    const n = Q[0].length;
    const Qt_b = [];
    for (let j = 0; j < n; j++) {
        const qj = getColumn(Q, j);
        Qt_b.push(dotProduct(qj, b));
    }

    // 解 Rx = Qᵀb
    return solveUpperTriangular(R, Qt_b);
}

// 矩陣乘法
function matrixMultiply(A, B) {
    const m = A.length;
    const k = B.length;
    const n = B[0].length;

    const result = Array.from({length: m}, () => Array(n).fill(0));
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            for (let p = 0; p < k; p++) {
                result[i][j] += A[i][p] * B[p][j];
            }
        }
    }
    return result;
}

// 矩陣轉置
function transpose(A) {
    const m = A.length;
    const n = A[0].length;
    return Array.from({length: n}, (_, j) =>
        Array.from({length: m}, (_, i) => A[i][j])
    );
}

function main() {
    printSeparator('QR 分解示範 (JavaScript)');

    // ========================================
    // 1. 基本 QR 分解
    // ========================================
    printSeparator('1. 基本 QR 分解');

    const A = [
        [1, 1],
        [1, 0],
        [0, 1]
    ];

    console.log('輸入矩陣 A：');
    printMatrix('A', A);

    const [Q, R] = qrDecomposition(A);

    console.log('\nQR 分解結果：');
    printMatrix('Q', Q);
    printMatrix('\nR', R);

    // 驗證 QᵀQ = I
    const QT = transpose(Q);
    const QTQ = matrixMultiply(QT, Q);
    console.log('\n驗證 QᵀQ = I：');
    printMatrix('QᵀQ', QTQ);

    // 驗證 A = QR
    const QR_result = matrixMultiply(Q, R);
    console.log('\n驗證 A = QR：');
    printMatrix('QR', QR_result);

    // ========================================
    // 2. 用 QR 解最小平方
    // ========================================
    printSeparator('2. 用 QR 解最小平方');

    // 數據
    const t = [0, 1, 2];
    const b = [1, 3, 4];

    console.log('數據點：');
    for (let i = 0; i < t.length; i++) {
        console.log(`  (${t[i]}, ${b[i]})`);
    }

    // 設計矩陣
    const A_ls = t.map(ti => [1, ti]);

    console.log('\n設計矩陣 A：');
    printMatrix('A', A_ls);
    printVector('觀測值 b', b);

    // QR 分解
    const [Q_ls, R_ls] = qrDecomposition(A_ls);
    printMatrix('\nQ', Q_ls);
    printMatrix('R', R_ls);

    // 解最小平方
    const x = qrLeastSquares(A_ls, b);
    printVector('\n解 x', x);

    console.log(`\n最佳直線：y = ${x[0].toFixed(4)} + ${x[1].toFixed(4)}t`);

    // ========================================
    // 3. 3×3 矩陣的 QR 分解
    // ========================================
    printSeparator('3. 3×3 矩陣的 QR 分解');

    const A2 = [
        [1, 1, 0],
        [1, 0, 1],
        [0, 1, 1]
    ];

    console.log('輸入矩陣 A：');
    printMatrix('A', A2);

    const [Q2, R2] = qrDecomposition(A2);

    console.log('\nQR 分解結果：');
    printMatrix('Q', Q2);
    printMatrix('\nR', R2);

    // 總結
    printSeparator('總結');
    console.log(`
QR 分解核心：

1. A = QR
   - Q: 標準正交矩陣 (QᵀQ = I)
   - R: 上三角矩陣

2. Gram-Schmidt 演算法：
   - 對 A 的行向量正交化得到 Q
   - R 的元素是投影係數

3. 用 QR 解最小平方：
   min ‖Ax - b‖²
   → Rx = Qᵀb

4. 優勢：
   - 比正規方程更穩定
   - 避免計算 AᵀA
`);

    console.log('='.repeat(60));
    console.log('示範完成！');
    console.log('='.repeat(60));
}

main();
